openmmtools/cache.py

```diff
@@ -387,8 +388,10 @@ def get_context(self, thermodynamic_state, integrator=None):
                 # Only one match.
                 context = self._lru[matching_context_ids[0]]
             else:
-                # Multiple matches, prefer non-default Integrator.
-                for context_id in matching_context_ids:
+                # Multiple matches, prefer the non-default Integrator.
+                # Always pick the least recently used to make two consective
+                # calls retrieving the same integrator.
+                for context_id in reversed(matching_context_ids):
                     if context_id[1] != self._default_integrator_id():
                         context = self._lru[context_id]
                         break
```

Comment:
Sorry for the confusion, the default integrator should actually be selected only if it's the only available option in the `elif` just above it.